src/testutils.rs

```diff
@@ -12,56 +12,65 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use std::fmt::{Debug, Error, Formatter};
-use std::io::Cursor;
 use std::path::{Path, PathBuf};
 
-use jujutsu_lib::testutils::{new_user_home, user_settings};
+use itertools::Itertools;
+use regex::Regex;
+use tempfile::TempDir;
 
-use crate::commands;
-use crate::ui::Ui;
-
-pub struct CommandRunner {
-    pub cwd: PathBuf,
-    pub stdout_buf: Vec<u8>,
+pub struct TestEnvironment {
+    _temp_dir: TempDir,
+    env_root: PathBuf,
+    home_dir: PathBuf,
 }
 
-impl CommandRunner {
-    pub fn new(cwd: &Path) -> CommandRunner {
-        CommandRunner {
-            cwd: cwd.to_owned(),
-            stdout_buf: vec![],
+impl Default for TestEnvironment {
+    fn default() -> Self {
+        let tmp_dir = TempDir::new().unwrap();
+        let env_root = tmp_dir.path().canonicalize().unwrap();
+        let home_dir = env_root.join("home");
+        Self {
+            _temp_dir: tmp_dir,
+            env_root,
+            home_dir,
         }
     }
+}
 
-    pub fn run(self, mut args: Vec<&str>) -> CommandOutput {
-        let _home_dir = new_user_home();
-        let mut stdout_buf = self.stdout_buf;
-        let stdout = Box::new(Cursor::new(&mut stdout_buf));
-        let ui = Ui::new(self.cwd, stdout, false, user_settings());
-        args.insert(0, "jj");
-        let status = commands::dispatch(ui, args);
-        CommandOutput { status, stdout_buf }
+impl TestEnvironment {
+    pub fn jj_cmd(&self, current_dir: &Path, args: &[&str]) -> assert_cmd::Command {
+        let mut cmd = assert_cmd::Command::cargo_bin("jj").unwrap();
+        cmd.current_dir(current_dir);
+        cmd.args(args);
+        cmd.env("HOME", self.home_dir.to_str().unwrap());
```

Comment:
When launching executables, you might might also want to call `.env_clear` and forward any other environment variables explicitly. git-branchless had a lot of issues where tests failed on various build environments/platforms.

(One frustrating example among many is the `GIT_INDEX_FILE` environment variable, which caused failures only 1. when running a pre-commit hook 2. as part of a `git commit --amend` 3. on tests which queried `git status` explicitly, rather than committing a known set of files. Hopefully jj has fewer integration issues!)

Reply:
> When launching executables, you might might also want to call `.env_clear` and forward any other environment variables explicitly. git-branchless had a lot of issues where tests failed on various build environments/platforms.

Oh, I had that but then I removed it because something made me think that it was just about clearing previously set overrides. But after looking at the docs again, I see that you seem to be right. Thanks!

> (One frustrating example among many is the `GIT_INDEX_FILE` environment variable, which caused failures only 1. when running a pre-commit hook 2. as part of a `git commit --amend` 3. on tests which queried `git status` explicitly, rather than committing a known set of files. Hopefully jj has fewer integration issues!)

Wow, that's weird :(